http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/LeoPublisher.scala

```diff
@@ -0,0 +1,60 @@
+package org.broadinstitute.dsde.workbench.leonardo
+
+import cats.effect.{Sync, Timer}
+import cats.implicits._
+import com.google.protobuf.ByteString
+import com.google.pubsub.v1.PubsubMessage
+import fs2.concurrent.InspectableQueue
+import fs2.{Pipe, Stream}
+import io.chrisdavenport.log4cats.Logger
+import io.circe.syntax._
+import org.broadinstitute.dsde.workbench.google2.GooglePublisher
+import org.broadinstitute.dsde.workbench.leonardo.db.{clusterQuery, DbReference}
+import org.broadinstitute.dsde.workbench.leonardo.http.{dbioToIO, nowInstant}
+import org.broadinstitute.dsde.workbench.leonardo.monitor.LeoPubsubCodec._
+import org.broadinstitute.dsde.workbench.leonardo.monitor.LeoPubsubMessage
+
+/**
+ * dequeue from publisherQueue and transform it into a native google pubsub message;
+ * After pubsub message is published, we update database when necessary
+ */
+final class LeoPublisher[F[_]: Logger: Timer](
+  publisherQueue: InspectableQueue[F, LeoPubsubMessage],
+  googlePublisher: GooglePublisher[F]
+)(implicit F: Sync[F], dbReference: DbReference[F]) {
+  val process: Stream[F, Unit] =
+    Stream.eval(Logger[F].info(s"Initializing publisher")) ++ publisherQueue.dequeue.flatMap { event =>
+      val publishing = Stream
+        .eval(F.pure(event))
+        .covary[F] through convertToPubsubMessagePipe through googlePublisher.publishNative
+      publishing.evalMap(_ => updateDatabase(event))
+    }
+
+  private def convertToPubsubMessagePipe: Pipe[F, LeoPubsubMessage, PubsubMessage] =
+    in =>
+      in.map { msg =>
+        val stringMessage = msg.asJson.noSpaces
+        val byteString = ByteString.copyFromUtf8(stringMessage)
+        PubsubMessage
+          .newBuilder()
+          .setData(byteString)
+          .putAttributes("traceId", msg.traceId.map(_.asString).getOrElse("null"))
+          .build()
+      }
+
+  private def updateDatabase(msg: LeoPubsubMessage): F[Unit] =
+    for {
+      now <- nowInstant[F]
+      _ <- msg match {
```

Comment:
Hm does this need to handle `LeoPubSubMessage.UpdateRuntimeMessage`?

Reply:
I don't see db update where `UpdateRuntimeMessage` is being used..maybe I'm missing something?

Reply:
Oh - `Updating` DB status is set in the `LeoPubSubMessageSubscriber`, because we only need to do it for certain types of updates.  This seems fine then.